File: leetcode_python/322.Coin_Change.py
#!/usr/bin/env python3
# coding: utf-8

# https://leetcode.com/problems/coin-change/
# 经典bottom-up， dp题目
# 子任务是 min(dp[i-coin_1], dp[i-coin_2], dp[i-coin_3) + 1

# Time complexity: O()
# Space complexity: O()

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        """
          0 1---------11  (amount)
        1  
        2 
        5 
        """
        dp = [0] + [amount + 1] * amount

        for i in range(1, len(dp)):
            for c in coins:
                if i >= c:
                    dp[i] = min(dp[i - c] + 1, dp[i])

        if dp[-1] == amount + 1:
            return -1
        else:
            return dp[-1]

File: leetcode_cpp/322.Coin_Change.cpp
// https://leetcode.com/problems/coin-change/

class Solution {
 public:
  int coinChange(vector<int>& coins, int amount) {
    vector<int> dp(amount + 1, amount + 1);  // (size, default_value)
    dp[0] = 0;

    for (int i = 1; i < amount + 1; i++) {
      for (auto c : coins) {
        if (i >= c) {
          dp[i] = min(dp[i], dp[i - c] + 1);
        }
      }
    }

    return dp[amount] == amount + 1 ? -1 : dp[amount];
  }
};